### DISCUSSION TITLE
GraphQL PriceAdjustment dynamic price update support

### DISCUSSION THREAD

**Discussion by michaelhaessig** [Shopify/function-examples#225] [Feedback: Cart Transform API]
This request is similar to #217 but i like to provide some more context to our use case.

In our apps we offer the ability to create dynamic price increases on a product.  For example offering an engraving option that adds 1$ to the price per letter entered. As price increases are dynamic, we cannot pre configure different variants and creating products on the fly also is not a valid option.  

Our app is currently using Draft orders with custom line items that have their own price set by our app when the customer checks out to add dynamic price additions.

I believe the Cart Transform API has the potential to finally support dynamic pricing (in a secure way).

In our tests with the new API we found that price of a Expand/Merge Operation can only be reduced via: [PriceAdjustment.percentageDecrease](https://shopify.dev/docs/api/functions/reference/cart-transform/graphql/common-objects/priceadjustment)

### Request
Make the API support setting a newly calculated price instead of increase/decrease only.  E.g:

````
"price": {
      "fixedAmount": {
           "value": 50
      }
}
````

Alternatively adding the ability to add a percentage based increase to the current decrease only feature would also work. But a custom price would certainly be a lot more flexible. 

### Final words
Adding price update support via Functions would enable lots more use cases. Maybe the Expand/Merge operations are not the right place, and a new Update operation would make more sense.  

There is a whole category of apps that currently use different hacks to support dynamic price update (using Draft Orders, custom product additions, dynamic variant creation). All of these requirements could be solved via a Shopify Function API that supports updating prices of existing products dynamically. 

I am very optimistic that the Function API approach is a great starting point for building a powerful checkout API, i'm looking forward to what the team will allow us to build on top of it!

**Comment by ignacio-chiazzo** [comment]
Thanks for the input!

> Adding price update support via Functions would enable lots more use cases. Maybe the Expand/Merge operations are not the right place, and a new Update operation would make more sense.

This makes sense and it's something we're considering. I'd like to understand more about your use case.

If we provided that API, how and where would you set the price to the variant considering that the function wouldn't have the ability to make a call to your system?.

**Reply by michaelhaessig** [reply]
Hi, awesome to hear that you guys are considering support for price updates!

To expand on our current solution; 
- On the product page our app can have many different options that then calculate an additional price on top of the base product price
- On add to cart, our app stores the total price increase on the line_item.properties
- On checkout, our app captures the full cart, creates a Draft order via our own API, which reads the line_item.proeprties and creates the additional custom line items for the add-on.   
- The checkout then shows additional line items for the price add-ons, which is not really user friendly (we get frequent support request for this). Other options apps all have the same problem.

Using line_item.properties avoids the need to call an external system, but obviously has some drawbacks, e.g. a bad actor could manually alter the line item properties to manipulate the pricing. However, this risk already exists with our current system based on draft orders. I believe it is the function's responsibility to validate whether a price modification is warranted. In our case, we only allow price increases, so this is not a concern.  (Based on thousands of orders we never had an issue where price add-ons were modified manually)

A fixedAmount api would allow the greatest flexibility,  but it might come in conflict with normal discounting practices. 
In that case a new "priceIncrease" API would also be sufficient for our use case.

**Reply by ignacio-chiazzo** [reply]
Thanks for the feedback:

> On checkout, our app captures the full cart, creates a Draft order via our own API, which reads the line_item.proeprties and creates the additional custom line items for the add-on.

It seems that you need the ability to add products to the cart and assign a price to support add-on. If you add the product via Cart API and then adjust the price via CartTransform, would that help?

> The checkout then shows additional line items for the price add-ons, which is not really user friendly (we get frequent support request for this). Other options apps all have the same problem.

Are the new products adds-on, or are they just fake products to store the price increase?


> Using line_item.properties avoids the need to call an external system, but obviously has some drawbacks, e.g. a bad actor could manually alter the line item properties to manipulate the pricing

Could the information be stored in a Metafield and be used inside the `cartTransform` function to determine the price? Does the price depend on a dynamic system outside Shopify?


> In that case a new "priceIncrease" API would also be sufficient for our use case.

Note that if we provide the API, the price would be "an override" of the price. Buyers will not see the old price, and if the price is lower, they will not see it as a discount.

**Reply by michaelhaessig** [reply]
> It seems that you need the ability to add products to the cart and assign a price to support add-on. If you add the product via Cart API and then adjust the price via CartTransform, would that help?

This would be the plan; normal add via cart API, then updating the price to include the add-on price with a CartTransform function.  

> Are the new products adds-on, or are they just fake products to store the price increase?

The add-ons are just fake products to simulate price increase, it would be desirable to not have extra line items.   (Our app also has actual add-on products features that already work with normal products linked).

> Could the information be stored in a Metafield and be used inside the cartTransform function to determine the price? Does the price depend on a dynamic system outside Shopify?

The add-on prices are very dynamic (many different price add-on features, like per letter, or even javascript formulas). So a Metafield is not the best approach (it would work by calling our apps own API) but it would be more efficient just reading the add-on price from the main line_item.properties 

> Note that if we provide the API, the price would be "an override" of the price. Buyers will not see the old price, and if the price is lower, they will not see it as a discount.

It would be desirable to simply show an updated price and nothing else.  This would be amazing if possible.

**Reply by michaelhaessig** [reply]
@ignacio-chiazzo any update on this feature?  

Based on the [Roadmap](https://shopify.dev/docs/apps/checkout#shopify-functions), the Cart Transform API is planned to be released in July, i would be curious to know if the suggested features will be added in this release. 

Would appreciate any more information!

**Reply by ignacio-chiazzo** [reply]
Thanks for the suggestions!

> i would be curious to know if the suggested features will be added in this release.

Fixed Price adjustment won't be included in the July version. 

> The add-ons are just fake products to simulate price increase, it would be desirable to not have extra line items. 

I definitely agree with this, we don't suggest adding extra products just to model a more dynamic price. 

Dynamic price is something we're actively investigating but we don't have yet a timeline.

**Reply by michaelhaessig** [reply]
> Fixed Price adjustment won't be included in the July version.

Got it, thanks for the update!

> Dynamic price is something we're actively investigating but we don't have yet a timeline.

Glad to hear that you are actively investigating this, feel free to reach out if you need any more context or feedback.

**Comment by bureauvk** [comment]
This would be amazing. Creating product variations on the fly and then injecting html code into the cart is so messy.

**Comment by leighs-hammer** [comment]
Yeah we would love to see extension in this space, we do this with draft orders and its hacky and prone to error. So would love to see further pricing strategies extended.

**Comment by n8cotoa** [comment]
@ignacio-chiazzo Hey there! Came across this while looking into Cart Transforms. Are there any updates on Fixed Price adjustments?

**Comment by stevekaran** [comment]
Bump... My customer has a unique use case where their primary selling strategy is bundles with a fixed price. So rather than applying a discount, or calculating a percentage decrease, the ability to set a fixed price (preferably the price of the parent bundle product as it will only have the default variant) would be very useful. Providing full flexibility of adjusting the item/cart price as part of the Cart Transform API would go a long way

**Reply by stevekaran** [reply]
Interestingly this _"kind of"_ works with an `expand` operation. For example, with `expand`, if the parent bundle product itself is added to the cart, then the price of the cart item is the price of the parent bundle product, not the sum of the items/variants in the bundle. On checkout, the price of the child items are automatically adjusted evenly so that sum of the child items equals the total the bundle parent price.

For example, if the bundle product of 3 items is $10, and each item individually is $4, the cart line item is not the sum of the 3 items ($12, $4 * 3), but rather the parent bundle product price ($10). On checkout, when the bundle is expanded, each item is automatically re-priced to $3.33, $3.33, $3.34, so that sum of the bundle items matches the bundle parent product price, without having to specify a `percentageDecrease`

Why I say _"kind of"_ works is that this is not exactly intuitive. When someone thinks of bundling/grouping products together, a `merge` operation makes more logical sense than an `expand` operation.

Thoughts? Ideas? Hopefully collectively we can come up with flexible solutions for developers and merchants to satisfy a wide variety of use cases.

**Comment by dsnvs** [comment]
New Cart Transform API developer preview addresses this:

https://shopify.dev/docs/api/functions/reference/cart-transform/developer-preview#example-update-operation

**Comment by PhilipBeauford** [comment]
I have a related issue regarding the updateOperation.
When I update a price down to zero the line item is then moved to the bottom of the list of cart items, instead of staying in the order it was added to the cart.

My client has a current legacy bundle builder that uses parent products with hidden price fields, and for now the bundle parent product price needs to be hidden as a work-around while we update the bundle logic for extensibility. I'm able to change price to 0 and 'hide' the price field by removing the 'free' label but the parent products are being pushed to the bottom of the item list, negating the use case.

Please advise if the line items can be adjusted to not be relocated to bottom of list after price updates or when updated to $0.

**Comment by stevekaran** [comment]
Currently as part of a `merge` or `expand` operation, we have to calculate the price percentage decrease to the value we want the bundle to be. Doing so renders price field of the parent bundle product useless. A merchant would set the price on the parent bundle product, just like any other product, and would expect to see that price appear in the cart, not the sum of the child variants.

It would be nice if Cart Transform could use the parent bundle price, then automatically re-price the child variants, just like it does when specifying the `percentageDecrease`

**Comment by maloguertin** [comment]
Is there any timeline as to when the price update might leave developer preview?

**Comment by maloguertin** [comment]
update ([UpdateOperation](https://shopify.dev/api/functions/reference/cart-transform/graphql/common-objects/updateoperation))
A cart line update operation. Only stores on the Shopify Plus plan can use apps with update operations.

Is this gonna be permanently gated to Shopify Plus?!

Edit:

If anyone wants to support trying to remove the Shopify Plus limitation:
https://community.shopify.com/c/partners-discussion/please-consider-removing-shopify-plus-for-cart-transform-update/m-p/2399190#M2701

**Reply by psppro26** [reply]
@maloguertin Nice post, you might post it in "shopify functions" section https://community.shopify.com/c/shopify-functions/bd-p/functions for a better visibility

**Comment by dnpg** [comment]
Just adding some comments after all the latest releases.

With Expand we have more flexibility on setting the price of the final "Customized Bundle". However in the Merge Operation we are still limited to the percentageDecrease value.

It would be great to have the option to set the fixedAmount

```
{
  "operations": [
    {
      "merge": {
        "cartLines": [
          {
            "cartLineId": "gid://shopify/CartLine/2",
            "quantity": 1
          },
          {
            "cartLineId": "gid://shopify/CartLine/3",
            "quantity": 1
          },
          {
            "cartLineId": "gid://shopify/CartLine/4",
            "quantity": 1
          }
        ],
        "parentVariantId": "gid://shopify/ProductVariant/123",
        "price": {
          "fixedAmount": {
            "value": 45.5
          }
        },
        "title": "Meal Kit",
        "image": {
          "url": "https://cdn.shopify.com/[...]/custom-image.png"
        }
      }
    }
  ]
}

```

Then the expected transformation should also edit the pricing on the CartLine items using the Weight Price Algorithm as per Expand Operation:
https://shopify.dev/docs/api/functions/reference/cart-transform#weight-price-algorithm
